src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+                        value.clone(),
+                        Err(e),
+                        &self.config,
+                    )
+                    .expect("value from error does not fail"),
+                )
+            }
+        };
+        self.cache.insert(key, value).await
+    }
+}
+
+impl<CR, Upstream, C> Debug for Request<CR, Upstream, C>
+where
+    CR: Send + Sync,
+    Upstream: Send + Sync,
+    C: Clock + Send + Sync,
+{
+    fn fmt(&self, _: &mut Formatter<'_>) -> Result<(), core::fmt::Error> {
+        todo!()
+    }
+}
+
+impl<CR, Upstream, C> GetResponse for Request<CR, Upstream, C>
+where
+    CR: Clone + ComposeRequest + Debug + Sync,
+    Upstream: SendRequest<CR> + Send + Sync + 'static,
+    C: Clock + Debug + Send + Sync + 'static,
+{
+    fn get_response(
+        &mut self,
+    ) -> Pin<
+        Box<
+            dyn Future<Output = Result<Message<Bytes>, Error>>
+                + Send
+                + Sync
+                + '_,
+        >,
+    > {
+        Box::pin(self.get_response_impl())
+    }
+}
+
+//------------ RequestState ---------------------------------------------------
+/// States of the state machine in get_response_impl
+enum RequestState {
+    /// Initial state, perform a cache lookup.
+    Init,
+
+    /// Wait for a response and insert the response in the cache.
+    GetResponse(Key, Box<dyn GetResponse + Send + Sync>),
+
+    /// Wait for a response but do not insert the response in the cache.
+    GetResponseNoCache(Box<dyn GetResponse + Send + Sync>),
+}
+
+//------------ Key ------------------------------------------------------------
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+/// The key for cache entries.
+struct Key {
+    /// DNS name in the request.
+    qname: Dname<Vec<u8>>,
+
+    /// The request class. Always IN at the moment.
+    qclass: Class,
+
+    /// The requested type.
+    qtype: Rtype,
+
+    /// Value of the AD flag.
+    ad: bool,
+
+    /// Value of the CD flag.
+    cd: bool,
+
+    /// Value of the DO flag.
+    dnssec_ok: bool,
+
+    /// Value of the RD flag.
+    rd: bool,
+}
+
+impl Key {
+    /// Create a new key object.
+    fn new<TDN>(
+        qname: TDN,
+        qclass: Class,
+        qtype: Rtype,
+        ad: bool,
+        cd: bool,
+        dnssec_ok: bool,
+        rd: bool,
+    ) -> Key
+    where
+        TDN: ToDname,
+    {
+        let mut qname = qname.to_dname().expect("to_dname should not fail");
+
+        // Make sure qname is canonical.
+        qname.make_canonical();
+
+        Self {
+            qname,
+            qclass,
+            qtype,
+            ad,
+            cd,
+            dnssec_ok,
+            rd,
+        }
+    }
+}
+
+//------------ Value ----------------------------------------------------------
+
+#[derive(Debug)]
+/// The value to be cached.
+struct Value<C>
+where
+    C: Clock + Send + Sync,
+{
+    /// Creation time of the cache entry.
+    creation: C::Instant,
```

Comment:
I have a slight preference for `created_at` and `valid_for` instead of `creation` and `validity`.